packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts

```diff
@@ -0,0 +1,228 @@
+	saveDraftDisabledButton: `${ panel } button.editor-post-saved-state.is-saved[aria-disabled="true"]`,
+	switchToDraftButton: `${ panel } button.editor-post-switch-to-draft`,
+
+	// Preview
+	mobilePreviewButton: `${ panel } a.editor-post-preview`,
+	desktopPreviewButton: `${ panel } button.block-editor-post-preview__button-toggle`,
+	desktopPreviewMenuItem: ( target: PreviewOptions ) =>
+		`button[role="menuitem"] span:text("${ target }")`,
+	previewPane: ( target: PreviewOptions ) => `.is-${ target.toLowerCase() }-preview`,
+
+	// Publish
+	publishButton: `${ panel } button.editor-post-publish-button__button[aria-disabled="false"]`,
+
+	// Editor settings
+	settingsButton: `${ panel } .interface-pinned-items > button:first-child`,
+};
+
+/**
+ * Represents an instance of the WordPress.com Editor's navigation sidebar.
+ * The component is available only in the Desktop viewport.
+ */
+export class EditorToolbarComponent {
+	private page: Page;
+	private frameLocator: FrameLocator;
+
+	/**
+	 * Constructs an instance of the component.
+	 *
+	 * @param {Page} page The underlying page.
+	 * @param {FrameLocator} frameLocator Locator of the editor iframe.
+	 */
+	constructor( page: Page, frameLocator: FrameLocator ) {
+		this.page = page;
+		this.frameLocator = frameLocator;
+	}
+
+	/* General helper */
+
+	/**
+	 * Given a selector, determines whether the target button/toggle is
+	 * in an expanded state.
+	 *
+	 * If the toggle is in the on state or otherwise in an expanded
+	 * state, this method will return true. Otherwise, false.
+	 *
+	 * @param {string} selector Target selector.
+	 * @returns {Promise<boolean>} True if target is in an expanded state. False otherwise.
+	 */
+	async targetIsOpen( selector: string ): Promise< boolean > {
+		const locator = this.frameLocator.locator( selector );
+		const pressed = await locator.getAttribute( 'aria-pressed' );
+		const expanded = await locator.getAttribute( 'aria-expanded' );
+		return pressed === 'true' || expanded === 'true';
+	}
+
+	/* Block Inserter */
+
+	/**
+	 * Opens the block inserter.
+	 */
+	async openBlockInserter(): Promise< void > {
+		if ( ! ( await this.targetIsOpen( selectors.blockInserterButton ) ) ) {
+			const locator = this.frameLocator.locator( selectors.blockInserterButton );
+			await locator.click();
+		}
+	}
+
+	/**
+	 * Closes the block inserter.
+	 */
+	async closeBlockInserter(): Promise< void > {
+		if ( await this.targetIsOpen( selectors.blockInserterButton ) ) {
+			const locator = this.frameLocator.locator( selectors.blockInserterButton );
+			await locator.click();
+		}
+	}
+
+	/* Draft */
+
+	/**
+	 * Clicks on the 'Save Draft' button on the editor.
+	 *
+	 * If the button cannot be clicked, the method short-circuits.
+	 */
+	async saveDraft(): Promise< void > {
+		const saveButtonLocator = this.frameLocator.locator( selectors.saveDraftButton );
+
+		if ( ! saveButtonLocator.waitFor( { timeout: 5 * 1000 } ) ) {
```

Comment:
I believe this throws when it times out, right? We might need a try/catch here instead.

Reply:
Good point. I will re-think how this part is structured, there _must_ be a better way to do this.

Reply:
I've wrapped the `waitFor` call around a `try/catch` for now.

I almost went back to using `locator.count`. My personal feeling is that `locator.count` is a reliable way to compute the number of elements that match a selector at any given time, as recall that's the point of `locator` API. The only caveat with `locator.count` is that if a page is not loaded fully, extrapolating from how behavior for other methods like `page.$$` have behaved we may not get the full picture, but in practice I have not observed this issue with `locator.count`.

Reply:
Fixed in 4122ed003d9348d80efb3ad926838b891533454b.